### DISCUSSION TITLE
How to implement testing with usage of ORM?

### DISCUSSION THREAD

**Discussion by Druags** [sqlalchemy/sqlalchemy#11257] [Usage Questions]
Hello, i have a trouble with understanding of how to create some tests for ORM models.

I have multiple files: 
1. models.py, it contains ORM models which are inherited from Base class
2. database.py, it contains Base class, which is inherited from DeclarativeBase class, also here creates engine and session factory
3. queries.py, it contains queries to database, in some queries i pass ORM class name as argument
4. config.py, it contains settings of database url
5. sql_main.py, it contains main function calls for create tables and to fill them by data

I tried to create separate file to write tests in it, but ran into a problem, where in some cases tables doesn't create because of wrong Base object in memory and some other errors, based on poor architecture of my program. So, what is the optimal practice of creating test for ORM models and functions which works with them? I'm a little bit confused and can't find a good solution by myself. Here is the repository with code: https://github.com/Druags/parser

**Comment by CaselIT** [comment]
Hi,

In general you would use a testing framework like pytest or similar to write your test.
That would allow you to create test cases and has utilities to aid with test.
You would then create an engine / session connected to a disposable db to use in the test, so that you use them to test your functionalities. This test engine/session are usually abstracted as fixtures so that you can re-use them in multiple tests
